query_optimizer/resolver/Resolver.cpp

```diff
@@ -596,12 +596,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
       *create_index_statement.relation_name(), nullptr /* reference_alias */);
 
-  const std::string index_name = create_index_statement.index_name()->value();
-  return L::CreateIndex::Create(input, index_name);
+  const std::string &index_name = create_index_statement.index_name()->value();
+
+  // Resolve attribute references
+  const PtrList<ParseAttribute> *index_attributes = create_index_statement.attribute_list();
+  const std::vector<E::AttributeReferencePtr> &relation_attributes =
+      input->getOutputAttributes();
+  std::vector<E::AttributeReferencePtr> resolved_attributes;
+  if (index_attributes == nullptr) {
+    // specify to build index on all the attributes, if no attribute was specified
+    for (const E::AttributeReferencePtr &relation_attribute : relation_attributes) {
+      resolved_attributes.emplace_back(relation_attribute);
+    }
+  } else {
+    // otherwise specify to build index on the attributes that were given
+    for (const ParseAttribute &index_attribute : *index_attributes) {
+      bool is_resolved = false;
+      for (const E::AttributeReferencePtr &relation_attribute : relation_attributes) {
+        const std::string &relation_attr_name = relation_attribute->attribute_name();
+        const std::string &index_attr_name = index_attribute.attr_name()->value();
+        if (relation_attr_name.compare(index_attr_name) == 0) {
+          is_resolved = true;
+          resolved_attributes.emplace_back(relation_attribute);
+          break;
+        }
+      }
+      if (!is_resolved) {
+        THROW_SQL_ERROR_AT(&index_attribute)<< "Attribute "<< index_attribute.attr_name()->value()
+        << " is undefined for the relation "<< create_index_statement.relation_name()->value();
+      }
+    }
+  }
+
+  // Resolve index properties
+  std::shared_ptr<const IndexSubBlockDescription> index_description_shared;
+  if (create_index_statement.getIndexProperties()->isIndexDescriptionValid()) {
+    // create a deep copy of the index description and pass its ownership to the shared ptr
+    std::unique_ptr<IndexSubBlockDescription> index_description(new IndexSubBlockDescription());
+    index_description->CopyFrom(*create_index_statement.getIndexProperties()->getIndexDescription());
+    index_description_shared.reset(index_description.release());
+    DCHECK(index_description_shared != nullptr);
+  } else {
+    if (create_index_statement.getIndexProperties()->getInvalidPropertyNode() != nullptr) {
+      // if exact location is known in the parser, the error is thrown at that specific node
+      THROW_SQL_ERROR_AT(create_index_statement.getIndexProperties()->getInvalidPropertyNode())
+        << create_index_statement.getIndexProperties()->getReasonForInvalidIndexDescription();
+    } else {
+      // else the error is thrown at the index name node
+      THROW_SQL_ERROR_AT(create_index_statement.index_type())
+        << create_index_statement.getIndexProperties()->getReasonForInvalidIndexDescription();
```

Comment:
FYI, you could simplify the whole `if-else` code fragment by keeping a reference to `create_index_statement.getIndexProperties()`.